src/aquarium-optimized/Aquarium.cpp

```diff
@@ -786,219 +782,112 @@ void Aquarium::render() {
-    model = mAquariumModels[i];
-    updateWorldMatrix(model);
-  }
-}
-
-void Aquarium::updateFishes() {
-  int begin =
+  int fishEnd =
       toggleBitset.test(static_cast<size_t>(TOGGLE::ENABLEINSTANCEDDRAWS))
-          ? MODELNAME::MODELSMALLFISHAINSTANCEDDRAWS
-          : MODELNAME::MODELSMALLFISHA;
-  int end = toggleBitset.test(static_cast<size_t>(TOGGLE::ENABLEINSTANCEDDRAWS))
-                ? MODELNAME::MODELBIGFISHBINSTANCEDDRAWS
-                : MODELNAME::MODELBIGFISHB;
-
-  for (int i = begin; i <= end; ++i) {
-    FishModel *model = static_cast<FishModel *>(mAquariumModels[i]);
-
-    const Fish &fishInfo = fishTable[i - begin];
-    int numFish = fishCount[i - begin];
-
-    model->prepareForDraw();
-
-    float fishBaseClock = g.mclock * g_fishSpeed;
-    float fishRadius = fishInfo.radius;
-    float fishRadiusRange = fishInfo.radiusRange;
-    float fishSpeed = fishInfo.speed;
-    float fishSpeedRange = fishInfo.speedRange;
-    float fishTailSpeed = fishInfo.tailSpeed * g_fishTailSpeed;
-    float fishOffset = g_fishOffset;
-    // float fishClockSpeed  = g_fishSpeed;
-    float fishHeight = g_fishHeight + fishInfo.heightOffset;
-    float fishHeightRange = g_fishHeightRange * fishInfo.heightRange;
-    float fishXClock = g_fishXClock;
-    float fishYClock = g_fishYClock;
-    float fishZClock = g_fishZClock;
-
-    for (int ii = 0; ii < numFish; ++ii) {
-      float fishClock = fishBaseClock + ii * fishOffset;
-      float speed = fishSpeed +
-                    static_cast<float>(matrix::pseudoRandom()) * fishSpeedRange;
-      float scale = 1.0f + static_cast<float>(matrix::pseudoRandom()) * 1;
-      float xRadius = fishRadius + static_cast<float>(matrix::pseudoRandom()) *
-                                       fishRadiusRange;
-      float yRadius =
-          2.0f + static_cast<float>(matrix::pseudoRandom()) * fishHeightRange;
-      float zRadius = fishRadius + static_cast<float>(matrix::pseudoRandom()) *
-                                       fishRadiusRange;
-      float fishSpeedClock = fishClock * speed;
-      float xClock = fishSpeedClock * fishXClock;
-      float yClock = fishSpeedClock * fishYClock;
-      float zClock = fishSpeedClock * fishZClock;
-
-      model->updateFishPerUniforms(
-          sin(xClock) * xRadius, sin(yClock) * yRadius + fishHeight,
-          cos(zClock) * zRadius, sin(xClock - 0.04f) * xRadius,
-          sin(yClock - 0.01f) * yRadius + fishHeight,
-          cos(zClock - 0.04f) * zRadius, scale,
-          fmod((g.mclock + ii * g_tailOffsetMult) * fishTailSpeed * speed,
-               static_cast<float>(M_PI) * 2),
-          ii);
-    }
-  }
+          ? MODELNAME::MODELBIGFISHBINSTANCEDDRAWS
+          : MODELNAME::MODELBIGFISHB;
 
-  mContext->updateAllFishData();
-}
+  for (int i = 0; i <= MODELNAME::MODELMAX; ++i) {
+    if (i >= MODELNAME::MODELSMALLFISHA && (i < fishBegin || i > fishEnd))
+      continue;
```

Comment:
Why

```c++
FOR_EACH_MEMBER(i) {
  if (IS_MEMBER_OF_X(i)) continue;
  if (IS_MEMBER_OF_Y(i)) { doSomething(i); }
  else /*IS_MEMBER_OF_Z(i)*/ { doSomethingElse(i); }
}
```

than

```c++
FOR_EACH_MEMBER_Y(i) { doSomething(i); }
FOR_EACH_MEMBER_Z(i) { doSomethingElse(i); }
```

Reply:
I hoped to keep the same structure of for loop. But your proposal is more clean, so I changed the code.